src/virtac/pv.py

```diff
@@ -0,0 +1,607 @@
+PytacItemType: TypeAlias = pytac.lattice.EpicsLattice | pytac.element.Element
+CallbackType: TypeAlias = Callable[[RecordValueType, int | None], None]
+
+
+class RecordTypes(StrEnum):
+    """Currently supported EPICS sofioc record types"""
+
+    AI = "ai"
+    AO = "ao"
+    WAVEFORM_OUT = "wfmo"
+    WAVEFORM_IN = "wfmi"
+    MBBI = "mbbi"
+
+
+@dataclass
+class RecordData:
+    """Class for holding information required to create a softioc record"""
+
+    record_type: str
+    lower: str | None = None
+    upper: str | None = None
+    precision: str | None = None
+    drive_high: str | None = None
+    drive_low: str | None = None
+    zrvl: str | None = None
+    zrst: str | None = None
+    scan: str = "I/O Intr"
+    pini: str = "YES"
+    always_update: bool = False
+    initial_value: RecordValueType = 0
+
+
+class BasePV:
+    """Stores the attributes and methods which allow the VIRTAC to control an
+    EPICS PV.
+
+
+    Attributes:
+        self.name (str): The name used to get both the PV and its softioc record.
+        self._record (softioc.pythonSoftIoc.RecordWrapper): This softioc record is the
+            heart of the PV class, the main purpose of PV objects is to manage the
+            setting and getting of these records.
+    """
+
+    def __init__(self, name: str, record_data: RecordData | None) -> None:
+        """
+        Args:
+            name (str): Used to identify this PV and its softioc record.
+            record_data (RecordData | None): Dataclass used to create this PVs softioc
+                record.
+        """
+        logging.debug(f"Creating PV: {name}")
+        self.name: str = name
+        self._record: RecordWrapper
+        if record_data is not None:
+            self.create_softioc_record(record_data)
+
+    def _on_update(self, value: RecordValueType, name: str) -> None:
+        """The callback function called when the softioc record updates.
+
+        This function and any overrides need to be kept FAST as they can be called
+        rapidly by CA clients.
+
+        Args:
+            value (RecordValueType): The value that has just been set to the record.
+            name (str): The name of the softioc record that has just been set to.
+        """
+        logging.debug("Read value %s on pv %s", value, name)
+
+    def set_record_field(self, field: str, value: RecordValueType) -> None:
+        """Set a field on this PVs softioc record.
+
+        Args:
+            field (softioc.field): The EPICS field to set on the softioc record
+            value (RecordValueType | str): The value to set to the EPICS field
+        """
+        self._record.set_field(field, value)
+
+    def create_softioc_record(
+        self,
+        record_data: RecordData,
+    ) -> None:
+        """Create this PVs softioc record.
+
+        Args:
+            record_data (RecordData): Dataclass used to create this PVs softioc record.
+        """
+        if "self._record" in locals():
```

Comment:
Perhaps this variable should be properly type hinted and initialised in the init?

Reply:
I have never seen this way of checking for an attribute before.